### PR TITLE
GenesisConfig for beacon client

### PR DESCRIPTION
**Pull request by claravanstaden**
The `initial_sync` extrinsic should be removed - it should form part of the pallet GenesisConfig.

### CODE REVIEW

File: parachain/pallets/ethereum-beacon-client/src/lib.rs [hunk 1]
```diff
@@ -712,7 +700,7 @@ pub mod pallet {
 			branch: Vec<H256>,
 			depth: u64,
 			index: u64,
-			root: Root,
+			root: H256,
```
Comment (claravanstaden) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:703]: I should revert this change back to `Root`.

File: parachain/primitives/beacon/src/lib.rs [hunk 2]
```diff
@@ -21,6 +29,47 @@ impl Default for PublicKey {
...
+}
+
+struct PublicKeyVisitor;
+
+#[cfg(feature = "std")]
+impl<'de> Visitor<'de> for PublicKeyVisitor {
+	type Value = PublicKey;
+
+	fn expecting(&self, formatter: &mut Formatter) -> StdResult {
+		formatter.write_str("an array of bytes")
```
Comment (claravanstaden) [inline parachain/primitives/beacon/src/lib.rs:46]:
```suggestion
		formatter.write_str("A hex string")
```

File: parachain/primitives/beacon/src/lib.rs [hunk 3]
```diff
@@ -21,6 +29,47 @@ impl Default for PublicKey {
...
+			None => v,
+		};
+
+		let hex_bytes = match hex::decode(str_without_0x) {
+			Ok(bytes)=> bytes,
+			Err(e)=> return Err(Error::custom(e.to_string()))
+		};
+
+		let mut data = [0u8; 48];
+		data[0..48].copy_from_slice(&hex_bytes);
```
Comment (claravanstaden) [inline parachain/primitives/beacon/src/lib.rs:61]: Need to check if `hex_bytes` length is `48` else error.

File: parachain/pallets/ethereum-beacon-client/src/lib.rs [hunk 4]
```diff
@@ -44,6 +47,7 @@ type ProofBranch = Vec<H256>;
 type ForkVersion = [u8; 4];
 
 #[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (vgeddes) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:50]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

Then we drop the serde imports on line 23.

File: parachain/primitives/beacon/src/lib.rs [hunk 5]
```diff
@@ -21,6 +28,47 @@ impl Default for PublicKey {
...
+
+#[cfg(feature = "std")]
+impl<'de> Visitor<'de> for PublicKeyVisitor {
+	type Value = PublicKey;
+
+	fn expecting(&self, formatter: &mut Formatter) -> StdResult {
+		formatter.write_str("an array of bytes")
+	}
+
+	fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>where E: Error,  {
```
Comment (vgeddes) [inline parachain/primitives/beacon/src/lib.rs:48]:
```suggestion
	fn visit_str<E>(self, v: &str) -> Result<Self::Value, E> where E: Error {	
```

File: parachain/primitives/beacon/src/lib.rs [hunk 6]
```diff
@@ -21,6 +28,47 @@ impl Default for PublicKey {
...
+#[cfg(feature = "std")]
+impl<'de> Visitor<'de> for PublicKeyVisitor {
+	type Value = PublicKey;
+
+	fn expecting(&self, formatter: &mut Formatter) -> StdResult {
+		formatter.write_str("an array of bytes")
+	}
+
+	fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>where E: Error,  {	
+        let	str_without_0x = match v.strip_prefix("0x") {
```
Comment (vgeddes) [inline parachain/primitives/beacon/src/lib.rs:49]:
```suggestion
        let str_without_0x = match v.strip_prefix("0x") {
```

File: relayer/relays/beacon/parachain-writer.go [hunk 7]
```diff
@@ -120,3 +120,18 @@ func (wr *ParachainWriter) WriteToParachain(ctx context.Context, extrinsicName s
...
+func (wr *ParachainWriter) getLastSyncedSyncCommitteePeriod() (uint64, error) {
+	key, err := types.CreateStorageKey(wr.conn.Metadata(), "EthereumBeaconClient", "LatestSyncCommitteePeriod", nil, nil)
+	if err != nil {
+		return 0, fmt.Errorf("create storage key for last sync committee: %w", err)
+	}
+
+	var period types.U64
+	ok, err := wr.conn.API().RPC.State.GetStorageLatest(key, &period)
+	if err != nil || !ok {
+		panic(err)
```
Comment (vgeddes) [inline relayer/relays/beacon/parachain-writer.go:133]: Can't we rather return an error here instead of panicking?

**Review by vgeddes — APPROVED** [review]
Looks good! Added some misc comments

File: parachain/pallets/ethereum-beacon-client/src/lib.rs [hunk 8]
```diff
@@ -792,6 +780,31 @@ pub mod pallet {
 				},
 			}
 		}
+
+		pub fn initial_sync(
```
Comment (alistair-singh) [inline parachain/pallets/ethereum-beacon-client/src/lib.rs:784]: Does this need to be `pub` if it is only called from `GenesisBuild::build`?

File: test/.envrc-example [hunk 9]
```diff
@@ -25,10 +25,11 @@ export PARACHAIN_ID=1000
...
+
 # export ROPSTEN_PRIVATE_KEY=         # Your deployment account private key
 # export BEEFY_RELAY_ETH_KEY=         # Your Beefy relayer account private key
 # export PARACHAIN_RELAY_ETH_KEY=     # Your Parachain relayer account private key
 # export E2E_TEST_ETH_KEY=            # Your E2E test account private key
 
-#export START_BEACON_SYNC=false       # Start the Beacon chain syncing (in development)
\ No newline at end of file
+# export START_BEACON_SYNC=false       # Start the Beacon chain syncing (in development)
+# export INITIAL_BEACON_BLOCK=0x088241fcf1cf63040b804498c945d3a6ae5484e65692483747fba5b8902c99c9 # Initial block to be synced from (this value is on ropsten)
```
Comment (alistair-singh) [inline test/.envrc-example:35]: Can we get the initial beacon block using a script? We do this with `beefy`.

File: parachain/pallets/ethereum-beacon-client/src/lib.rs (hunk 8 above)
Comment (claravanstaden) [reply parachain/pallets/ethereum-beacon-client/src/lib.rs]: Good point! I changed it to `pub(super)` so that the tests can still access it.

File: test/.envrc-example (hunk 9 above)
Comment (claravanstaden) [reply test/.envrc-example:35]: We could just get the latest block, yeah. However, the light client protocol suggests always syncing from a manually validated block, so that you know you are syncing on the right chain. Is that not also an issue with Beefy?

Comment (alistair-singh) [reply test/.envrc-example:35]: Sorry maybe beefy is the wrong example here. In the `ethereum-light-client` we know the block is trusted by getting the latest block from a trusted node (Infura) as opposed to setting an environment variable. Can we not do something similar for beacon? Fetch the latest sync committee from a trusted beacon node at time of building the genesis state?

Comment (claravanstaden) [reply test/.envrc-example:35]: I can get the latest finalized block from a trusted node, yeah! Let me do that quick.

I think the sync committee fetching can stay in the relayer for now - I'll relook at it once I get to the [resuming syncing ticket](https://linear.app/snowfork/issue/SNO-292/relayer-and-beacon-client-should-support-syncing-from-last-synced-slot).

Comment (claravanstaden) [reply test/.envrc-example]: Added in https://github.com/Snowfork/snowbridge/pull/658/commits/31bb446bed0110d837aaa3430b498ca1fccb08bb.

**Review by alistair-singh — APPROVED** [review]
